modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/upgrade/v3_2_1/UpgradeDLFileEntryType.java

```diff
@@ -0,0 +1,240 @@
+					rs.getLong("groupId"), rs.getLong("companyId"),
+					rs.getLong("userId"), rs.getString("userName"),
+					rs.getString("name"), ddmStructureId,
+					ddmStructureVersionId);
+
+				ResourceLocalServiceUtil.addResources(
+					rs.getLong("companyId"), rs.getLong("groupId"),
+					rs.getLong("userId"),
+					ResourceActionsUtil.getCompositeModelName(
+						DLFileEntryMetadata.class.getName(),
+						DDMStructure.class.getName()),
+					ddmStructureId, false, false, false);
+
+				ps2.setLong(1, ddmStructureId);
+
+				ps2.setLong(2, rs.getLong("fileEntryTypeId"));
+
+				ps2.addBatch();
+			}
+
+			ps2.executeBatch();
+		}
+
+		String template = StringUtil.read(
+			UpgradeDLFileEntryType.class.getResourceAsStream(
+				"dependencies/update.sql"));
+
+		runSQLTemplateString(template, false);
+	}
+
+	private long _insertDDMStructure(
+			long groupId, long companyId, long userId, String userName,
+			String name)
+		throws Exception {
+
+		try (PreparedStatement ps2 = connection.prepareStatement(
+				"INSERT INTO DDMStructure VALUES (1,0,?,?,?,?,?,?,?,?,?" +
+					",?,0,?,?,'1.0',?,'',?,'default',1,NULL)")) {
+
+			long classNameId = PortalUtil.getClassNameId(
+				DLFileEntryMetadata.class);
+
+			ps2.setString(1, PortalUUIDUtil.generate());
+
+			long ddmStructureId = increment();
+
+			ps2.setLong(2, ddmStructureId);
+
+			ps2.setLong(3, groupId);
+			ps2.setLong(4, companyId);
+			ps2.setLong(5, userId);
+			ps2.setString(6, userName);
+			ps2.setLong(7, userId);
+			ps2.setString(8, userName);
+
+			Date now = new Date(System.currentTimeMillis());
+
+			ps2.setDate(9, now);
+			ps2.setDate(10, now);
+
+			ps2.setLong(11, classNameId);
+
+			ps2.setString(12, String.valueOf(ddmStructureId));
+			ps2.setString(13, name);
+			ps2.setString(
+				14,
+				StringUtil.read(
+					UpgradeDLFileEntryType.class.getResourceAsStream(
+						"dependencies/ddm-structure.json")));
+
+			ps2.executeUpdate();
+
+			return ddmStructureId;
+		}
+	}
+
+	private void _insertDDMStructureLayout(
+			long groupId, long companyId, long userId, String userName,
+			String name, long ddmStructureId, long ddmStructureVersionId)
+		throws Exception {
+
+		try (PreparedStatement ps = connection.prepareStatement(
+				"INSERT INTO DDMStructureLayout VALUES (0,0,?,?,?,?,?,?,?,?,?" +
+					",?,?,?,'',?)")) {
+
+			long classNameId = PortalUtil.getClassNameId(
+				DLFileEntryMetadata.class);
+
+			ps.setString(1, PortalUUIDUtil.generate());
+			ps.setLong(2, increment());
+			ps.setLong(3, groupId);
+			ps.setLong(4, companyId);
+			ps.setLong(5, userId);
+			ps.setString(6, userName);
+
+			Date now = new Date(System.currentTimeMillis());
+
+			ps.setDate(7, now);
+			ps.setDate(8, now);
+
+			ps.setLong(9, classNameId);
+
+			ps.setString(10, String.valueOf(ddmStructureId));
+			ps.setLong(11, ddmStructureVersionId);
+			ps.setString(12, name);
+			ps.setString(
+				13,
+				StringUtil.read(
+					UpgradeDLFileEntryType.class.getResourceAsStream(
```

Comment:
Can we read this file once, when instantiating the upgrade? This method is called from a loop, so this could be problematic.